### DISCUSSION TITLE
gRPC Metadata support

### DISCUSSION THREAD

**Discussion by Ten0** [Kong/insomnia#3667] [Ideas] [answered]
**Is your feature request related to a problem? Please describe.**
Can't use insomnia for gRPC where we pass data to some gRPC calls using metadata (e.g. authentication).
This is a [known lacking feature](https://support.insomnia.rest/article/188-grpc), but looks like it doesn't have a tracking issue yet.

**Describe the solution you'd like**
Being able to configure, by call, ideally the same exact interface as for GraphQL "query variables", the metadata to send.

This is where this would hit in the code:
https://github.com/Kong/insomnia/blob/f27cd5f6306eb50787e31d1f4cfe2606e0c03eb2/packages/insomnia-app/app/network/grpc/index.js#L52-L58

becoming in effect (as per the doc of  [`makeUnaryRequest`](https://grpc.github.io/grpc/node/grpc.Client.html#makeUnaryRequest__anchor)):
```typescript
let metaData = new grpc.Metadata();
for (const [key, value] of Object.entries(metaDataObj)) {
    metaData.add(key, value);
}
return client.makeUnaryRequest(
    path,
    requestSerialize,
    responseDeserialize,
    messageBody,
    metadata,
    callback,
  );
```

(and obviously same for the 3 other types of streaming)

**Comment by filfreire** [comment, marked as answer]
@Ten0 tiny update - we’ve started developing gRPC headers support, it’s still in progress, but you can follow progress here: https://github.com/Kong/insomnia/pull/4244

Edit: We have gRPC headers support since [2021.7.2](https://github.com/Kong/insomnia/releases/tag/core%402021.7.2) release.

**Reply by filfreire** [reply]
@Ten0 you should be able to try it out with the latest beta - https://github.com/Kong/insomnia/releases/tag/core%402021.7.0-beta.0
